sdk/eventhubs/azure-messaging-eventhubs/src/producer_client.cpp

```diff
@@ -169,22 +105,109 @@ namespace Azure { namespace Messaging { namespace EventHubs {
     Send(batch, context);
   }
 
+  Azure::Core::Amqp::_internal::Connection ProducerClient::CreateConnection()
+  {
+    Azure::Core::Amqp::_internal::ConnectionOptions connectOptions;
+    connectOptions.ContainerId = m_producerClientOptions.ApplicationID;
+    connectOptions.EnableTrace = _detail::EnableAmqpTrace;
+    connectOptions.AuthenticationScopes = {"https://eventhubs.azure.net/.default"};
+
+    // Set the UserAgent related properties on this message sender.
+    _detail::EventHubsUtilities::SetUserAgent(
+        connectOptions, m_producerClientOptions.ApplicationID);
+
+    return Azure::Core::Amqp::_internal::Connection{
+        m_fullyQualifiedNamespace, m_credential, connectOptions};
+  }
+  void ProducerClient::EnsureConnection(std::string const& partitionId)
+  {
+    std::unique_lock<std::mutex> lock(m_sessionsLock);
+    if (m_connections.find(partitionId) == m_connections.end())
+    {
+      m_connections.emplace(partitionId, CreateConnection());
+    }
+  }
+
+  void ProducerClient::EnsureSession(std::string const& partitionId)
+  {
+    // Ensure that a connection has been created for this producer.
+    EnsureConnection(partitionId);
+
+    // Ensure that a session has been created for this partition.
+    std::unique_lock<std::mutex> lock(m_sessionsLock);
+    if (m_sessions.find(partitionId) == m_sessions.end())
+    {
+      //      Azure::Core::Amqp::_internal::Session session{CreateSession()};
+      //      m_sessions.emplace(partitionId, std::move(session));
+      m_sessions.emplace(partitionId, CreateSession(partitionId));
+    }
+  }
+
+  Azure::Core::Amqp::_internal::Session ProducerClient::GetSession(std::string const& partitionId)
+  {
+    return m_sessions.at(partitionId);
```

Comment:
This needs to also go through mutex, and I think it would benefit from the `shared_lock`.
Please review the code for all the stuff that you access via mutexes for instances like this, if we ever access them, even for reading, without taking a mutex, and if it is safe.